### ISSUE TITLE
Add an API to discover if an attribute is a reserved attribute

### ISSUE THREAD

**Issue by ericstj** [dotnet/roslyn#68968]
## Background and Motivation

When examining symbols it would be helpful to understand if an attribute is reserved and only meant to be used by the compiler.

## Proposed API
```diff
namespace Microsoft.CodeAnalysis
{
     public class AttributeData
     {
+        public bool IsReserved { get; }
     }
```

## Usage Examples
Omitting reserved attributes when dumping attributes applied to a member.
``` C#
foreach (var attribute in symbol.GetAttributes())
{
    if (!attribute.IsReserved}
    {
        WriteAttribute(attribute);
    }
}
```

## Alternative Designs
We can hardcode our own list, but it will get out of date as new language features create new attributes.

**Comment by 333fred** [comment]
Many of these types of attributes are already hidden by the compiler: for example, `RequiredMemberAttribute` is removed in favor of `IsRequired`. What attributes are we concerned about here, and what's the scenario you're looking to display these in @ericstj?

Some other questions:

* Is this something that you only care about from metadata, only from source, or both?
* How would a metadata symbol vary across compilations? There are some attributes that are allowed in VB that are not in C#, for example, and vice versa. What should happen when metadata is viewed from the other side?

**Comment by ericstj** [comment]
We're examining metadata using ISymbol* and want to create a C# representation: https://github.com/dotnet/sdk/tree/main/src/GenAPI.  We use SyntaxGenerator for most of the work but have to handle attributes ourselves.  We'd like to avoid hardcoding which attributes to ignore.  See https://github.com/dotnet/sdk/pull/33888

**Comment by CyrusNajmabadi** [comment]
Sure.  makes sense for us to handle it.  but we'd be in teh same boat of asking for which APIs are things we shoudl filter out (unless you've already done that for us at the sybmolic layer).

**Comment by 333fred** [comment]
> but we'd be in teh same boat of asking for which APIs are things we shoudl filter out (unless you've already done that for us at the sybmolic layer).

Fair enough, there are still some questions to answer. Also cc @AlekseyTs as API owner.

**Comment by AlekseyTs** [comment]
If this is a real issue for source generators, I am leaning towards filtering out an attribute synthesized by compiler during import if the attribute cannot be applied in source at the same position and there is an alternative way to observe its presence (a special property on the symbol, etc. ). I know that compiler might prevent some attributes to be applied at some other positions (where compiler never synthesizes them), but I do not think we should be filtering them out at those positions.
